libbeat/publisher/queue/diskqueue/queue.go

```diff
@@ -0,0 +1,249 @@
+	"github.com/elastic/beats/v7/libbeat/logp"
+	"github.com/elastic/beats/v7/libbeat/publisher/queue"
+)
+
+// diskQueue is the internal type representing a disk-based implementation
+// of queue.Queue.
+type diskQueue struct {
+	logger   *logp.Logger
+	settings Settings
+
+	// Metadata related to the segment files.
+	segments *diskQueueSegments
+
+	// Metadata related to consumer acks / positions of the oldest remaining
+	// frame.
+	acks *diskQueueACKs
+
+	// The queue's helper loops, each of which is run in its own goroutine.
+	readerLoop  *readerLoop
+	writerLoop  *writerLoop
+	deleterLoop *deleterLoop
+
+	// Wait group for shutdown of the goroutines associated with this queue:
+	// reader loop, writer loop, deleter loop, and core loop (diskQueue.run()).
+	waitGroup sync.WaitGroup
+
+	// writing is true if the writer loop is processing a request, false
+	// otherwise.
+	writing bool
+
+	// reading is true if the reader loop is processing a request, false
+	// otherwise.
+	reading bool
+
+	// deleting is true if the deleter loop is processing a request, false
+	// otherwise.
+	deleting bool
+
+	// The API channel used by diskQueueProducer to write events.
+	producerWriteRequestChan chan producerWriteRequest
+
+	// pendingFrames is a list of all incoming data frames that have been
+	// accepted by the queue and are waiting to be sent to the writer loop.
+	// Segment ids in this list always appear in sorted order, even between
+	// requests (that is, a frame added to this list always has segment id
+	// at least as high as every previous frame that has ever been added).
+	pendingFrames []segmentedFrame
+
+	// blockedProducers is a list of all producer write requests that are
+	// waiting for free space in the queue.
+	blockedProducers []producerWriteRequest
+
+	// The channel to signal our goroutines to shut down.
+	done chan struct{}
+}
+
+func init() {
+	queue.RegisterQueueType(
+		"disk",
+		queueFactory,
+		feature.MakeDetails(
+			"Disk queue",
+			"Buffer events on disk before sending to the output.",
+			feature.Beta))
+}
+
+// queueFactory matches the queue.Factory interface, and is used to add the
+// disk queue to the registry.
+func queueFactory(
+	ackListener queue.ACKListener, logger *logp.Logger, cfg *common.Config,
+) (queue.Queue, error) {
+	settings, err := SettingsForUserConfig(cfg)
+	if err != nil {
+		return nil, fmt.Errorf("Disk queue couldn't load user config: %w", err)
+	}
+	settings.WriteToDiskListener = ackListener
+	return NewQueue(logger, settings)
+}
+
+// NewQueue returns a disk-based queue configured with the given logger
+// and settings, creating it if it doesn't exist.
+func NewQueue(logger *logp.Logger, settings Settings) (queue.Queue, error) {
+	logger = logger.Named("diskqueue")
+	logger.Debugf(
+		"Initializing disk queue at path %v", settings.directoryPath())
+
+	if settings.MaxBufferSize > 0 &&
+		settings.MaxBufferSize < settings.MaxSegmentSize*2 {
+		return nil, fmt.Errorf(
```

Comment:
Your error messages in `fmt.Errorf` calls should not start with capitalized letters.